pyfr/plugins/ascent.py

```diff
@@ -0,0 +1,415 @@
+        # Determine what gradients, if any, are required
+        g_pnames = set()
+        for expr in self._exprs.values():
+            if not isinstance(expr, dict):
+                g_pnames.update(re.findall(r'\bgrad_(.+?)_[xyz]\b', expr))
+            else:
+                for sexpr in expr.values():
+                    g_pnames.update(re.findall(r'\bgrad_(.+?)_[xyz]\b', sexpr))
+
+        privarmap = self.elementscls.privarmap[self.ndims]
+        self._gradpinfo = [(pname, privarmap.index(pname)) 
+                           for pname in g_pnames]
+    
+    def _init_pipelines(self, cfgsect):
+        pipel_cfg = {}
+        self.pipelines = pl = ConduitNode(self.conduit)
+
+        for k in self.cfg.items(cfgsect, prefix='pipeline-'):
+            pn = k.removeprefix('pipeline-')
+            cfg = self.cfg.getliteral(cfgsect, k)
+            pipel_cfg[pn] = cfg = [cfg] if isinstance(cfg, dict) else cfg
+
+            for j, filter in enumerate(cfg):
+                params = ConduitNode(self.conduit)
+
+                pl[f'pl_{pn}/f{j}/type'] = ptype = filter.pop('type')
+                for kf, vf in filter.items():
+                    if kf == 'output_name':
+                        if vf in self._fields_write:
+                            raise KeyError(f"Output name '{vf}' already used")
+                        else:
+                            self._fields_write.add(vf)
+                    elif kf == 'field':
+                        self._fields_read.add(vf)
+
+                    params[kf] = vf
+                
+                if ptype in self.v_filter:
+                    self._fields_read.add('velocity')
+
+                pl[f'pl_{pn}/f{j}/params'] = params
+
+    def _init_scenes(self, cfgsect):
+        cons = self.cfg.items_as('constants', float)
+        self._scene_cfg = scene_cfg = {}
+        self.scenes = ConduitNode(self.conduit)
+
+        for k in self.cfg.items(cfgsect, prefix='scene-'):
+            sn = k.removeprefix('scene-')
+            scene_cfg[sn] = cfg = self.cfg.getliteral(cfgsect, k)
+
+            if cfg['type'] != 'mesh':
+                field = cfg['field']
+                self._fields_read.add(field)
+
+            render_setting = False
+            for kc, vc in cfg.items():
+                if kc.startswith('field_expr'):
+                    self._add_field_expr(field, kc, vc, cons)
+                elif kc == 'pipeline':
+                    self.scenes[f's_{sn}/plots/p1/pipeline'] = f'pl_{vc}'
+                elif kc.startswith('render-'):
+                    ropt = kc.removeprefix('render-').replace('-', '/')
+                    self.scenes[f's_{sn}/renders/r1/{ropt}'] = vc
+                    render_setting = True
+                elif kc != 'image_name':
+                    self.scenes[f's_{sn}/plots/p1/{kc}'] = vc
+
+            # If render options then move name so used for render
+            rpath = 'renders/r1/' if render_setting else ''
```

Comment:
Might be cleaner to set the path with an `if any(kc.startswith('render-') for fc in cfg):` to avoid the variable.